src/pathfinder/water_regions.cpp

```diff
@@ -0,0 +1,405 @@
+	 */
+	TWaterRegionTraversabilityBits GetEdgeTraversabilityBits(DiagDirection side) const { return edge_traversability_bits[side]; }
+
+	/**
+	 * @returns The amount of individual water patches present within the water region. A value of
+	 * 0 means there is no water present in the water region at all.
+	 */
+	int NumberOfPatches() const { return this->number_of_patches; }
+
+	/**
+	 * @returns Whether the water region contains aqueducts that cross the region boundaries.
+	 */
+	bool HasCrossRegionAqueducts() const { return this->has_cross_region_aqueducts; }
+
+	/**
+	 * Returns the patch label that was assigned to the tile.
+	 * @param tile The tile of which we want to retrieve the label.
+	 * @returns The label assigned to the tile.
+	 */
+	TWaterRegionPatchLabel GetLabel(TileIndex tile) const
+	{
+		assert(this->tile_area.Contains(tile));
+		return this->tile_patch_labels[GetLocalIndex(tile)];
+	}
+
+	/**
+	 * Performs the connected component labeling and other data gathering.
+	 * @see WaterRegion
+	 */
+	void ForceUpdate()
+	{
+		this->has_cross_region_aqueducts = false;
+
+		std::unordered_map<int, TWaterRegionPatchLabel> temp_labels;
+		for (const TileIndex tile : this->tile_area) {
+			this->tile_patch_labels[GetLocalIndex(tile)] = INVALID_WATER_REGION_PATCH;
+			temp_labels[tile] = INVALID_WATER_REGION_PATCH;
+			if (IsAqueductTile(tile)) {
+				const TileIndex other_aqueduct_end = GetOtherBridgeEnd(tile);
+				if (!tile_area.Contains(other_aqueduct_end)) this->has_cross_region_aqueducts = true;
+			}
+		}
+
+		TWaterRegionPatchLabel current_label = 1;
+		TWaterRegionPatchLabel highest_assigned_label = 0;
+
+		/* Perform connected component labeling. This uses a flooding algorithm that expands until no
+		 * additional tiles can be added. Only tiles inside the water region are considered. */
+		for (const TileIndex start_tile : tile_area) {
+			static std::vector<TileIndex> tiles_to_check;
+			tiles_to_check.clear();
+			tiles_to_check.push_back(start_tile);
+
+			bool increase_label = false;
+			while (!tiles_to_check.empty()) {
+				const TileIndex tile = tiles_to_check.back();
+				tiles_to_check.pop_back();
+
+				const TrackdirBits valid_dirs = TrackBitsToTrackdirBits(GetWaterTracks(tile));
+				if (!valid_dirs) continue;
+
+				const TWaterRegionPatchLabel tile_label = temp_labels[tile];
+				if (tile_label != INVALID_WATER_REGION_PATCH) continue;
+
+				temp_labels[tile] = current_label;
+				highest_assigned_label = current_label;
+				increase_label = true;
+
+				for (const Trackdir dir : SetTrackdirBitIterator(valid_dirs)) {
+					/* By using a TrackFollower we "play by the same rules" as the actual ship pathfinder */
+					CFollowTrackWater ft;
+					if (ft.Follow(tile, dir) && this->tile_area.Contains(ft.m_new_tile)) tiles_to_check.push_back(ft.m_new_tile);
+				}
+			}
+
+			if (increase_label) current_label++;
+		}
+		for (const auto [tile, label] : temp_labels) tile_patch_labels[GetLocalIndex(tile)] = label;
```

Comment:
When multiple tiles within the same local index have a different labels, then this is nondeterministic and a desync threat.

Reply:
Looks like I don't really need the temp_labels anymore, I can directly write to the std::array instead with some modifications. It's cleaner too.